File: xue-rs/src/aggregator.rs
use colored::*;
use dashmap::DashMap;
use serde::Serialize;
use std::sync::atomic::{AtomicU64, AtomicUsize, Ordering};
use std::sync::Mutex;
use std::time::Instant;
use std::collections::HashMap;
use url::Url;

/// Thread-safe aggregator using atomic counters and DashMap.
/// No heavy locks — designed for high-concurrency access.
pub struct Aggregator {
    start_time: Instant,

    // Domain tracking (lock-free concurrent map)
    pub domains: DashMap<String, usize>,

    // Status code distribution
    pub status_codes: DashMap<u16, usize>,

    // Content types
    pub content_types: DashMap<String, usize>,

    // Atomic counters (no locks needed)
    pub onion_count: AtomicUsize,
    pub clearnet_count: AtomicUsize,
    pub total_links_found: AtomicUsize,
    pub total_size_bytes: AtomicU64,
    pub total_errors: AtomicUsize,

    // Error tracking (bounded, needs a lock)
    pub errors: DashMap<String, usize>,
    pub error_urls: Mutex<Vec<(String, String)>>,

    // Largest pages
    pub largest_pages: Mutex<Vec<(String, u64)>>,
}

impl Aggregator {
    pub fn new() -> Self {
        Self {
            start_time: Instant::now(),
            domains: DashMap::new(),
            status_codes: DashMap::new(),
            content_types: DashMap::new(),
            onion_count: AtomicUsize::new(0),
            clearnet_count: AtomicUsize::new(0),
            total_links_found: AtomicUsize::new(0),
            total_size_bytes: AtomicU64::new(0),
            total_errors: AtomicUsize::new(0),
            errors: DashMap::new(),
            error_urls: Mutex::new(Vec::new()),
            largest_pages: Mutex::new(Vec::new()),
        }
    }

    pub fn record_page(
        &self,
        url: &str,
        status_code: u16,
        content_type: &str,
        links_found: usize,
        size_bytes: u64,
    ) {
        // Domain
        if let Ok(parsed) = Url::parse(url) {
            if let Some(host) = parsed.host_str() {
                *self.domains.entry(host.to_string()).or_insert(0) += 1;

                if host.contains(".onion") {
                    self.onion_count.fetch_add(1, Ordering::Relaxed);
                } else {
                    self.clearnet_count.fetch_add(1, Ordering::Relaxed);
                }
            }
        }

        // Status code
        *self.status_codes.entry(status_code).or_insert(0) += 1;

        // Content type
        let ct = content_type.split(';').next().unwrap_or("unknown").trim().to_string();
        *self.content_types.entry(ct).or_insert(0) += 1;

        // Counters
        self.total_links_found.fetch_add(links_found, Ordering::Relaxed);
        self.total_size_bytes.fetch_add(size_bytes, Ordering::Relaxed);

        // Largest pages
        if let Ok(mut pages) = self.largest_pages.lock() {
            pages.push((url.to_string(), size_bytes));
            if pages.len() > 20 {
                pages.sort_by(|a, b| b.1.cmp(&a.1));
                pages.truncate(10);
            }
        }
    }

    pub fn record_error(&self, url: &str, error: &str) {
        self.total_errors.fetch_add(1, Ordering::Relaxed);

        let error_type = error.split(':').next().unwrap_or("Unknown").to_string();
        *self.errors.entry(error_type).or_insert(0) += 1;

        if let Ok(mut urls) = self.error_urls.lock() {
            urls.push((url.to_string(), error.to_string()));
            if urls.len() > 200 {
                urls.drain(0..100);
            }
        }
    }

    pub fn total_pages(&self) -> usize {
        self.onion_count.load(Ordering::Relaxed) + self.clearnet_count.load(Ordering::Relaxed)
    }

    pub fn total_domains(&self) -> usize {
        self.domains.len()
    }

    pub fn elapsed_secs(&self) -> f64 {
        self.start_time.elapsed().as_secs_f64()
    }

    pub fn pages_per_sec(&self) -> f64 {
        let elapsed = self.elapsed_secs();
        if elapsed > 0.0 {
            self.total_pages() as f64 / elapsed
        } else {
            0.0
        }
    }

    pub fn format_elapsed(&self) -> String {
        let secs = self.start_time.elapsed().as_secs();
        let h = secs / 3600;
        let m = (secs % 3600) / 60;
        let s = secs % 60;
        if h > 0 {
            format!("{}:{:02}:{:02}", h, m, s)
        } else {
            format!("{}:{:02}", m, s)
        }
    }

    pub fn live_stats_line(&self) -> String {
        format!(
            "{} crawled / {} domains / {:.1}/s / {}",
            self.total_pages(),
            self.total_domains(),
            self.pages_per_sec(),
            self.format_elapsed(),
        )
    }

    pub fn generate_report(&self) -> Report {
        let total = self.total_pages();
        let elapsed = self.elapsed_secs();

        let mut top_domains: Vec<(String, usize)> = self
            .domains
            .iter()
            .map(|e| (e.key().clone(), *e.value()))
            .collect();
        top_domains.sort_by(|a, b| b.1.cmp(&a.1));
        top_domains.truncate(30);

        let mut status_dist: Vec<(u16, usize)> = self
            .status_codes
            .iter()
            .map(|e| (*e.key(), *e.value()))
            .collect();
        status_dist.sort_by_key(|e| e.0);

        let mut ct_dist: Vec<(String, usize)> = self
            .content_types
            .iter()
            .map(|e| (e.key().clone(), *e.value()))
            .collect();
        ct_dist.sort_by(|a, b| b.1.cmp(&a.1));
        ct_dist.truncate(10);

        let mut error_dist: Vec<(String, usize)> = self
            .errors
            .iter()
            .map(|e| (e.key().clone(), *e.value()))
            .collect();
        error_dist.sort_by(|a, b| b.1.cmp(&a.1));

        let largest = self
            .largest_pages
            .lock()
            .map(|p| {
                let mut sorted = p.clone();
                sorted.sort_by(|a, b| b.1.cmp(&a.1));
                sorted.truncate(10);
                sorted
            })
            .unwrap_or_default();

        Report {
            summary: ReportSummary {
                total_pages_crawled: total,
                total_unique_domains: self.total_domains(),
                total_errors: self.total_errors.load(Ordering::Relaxed),
                onion_pages: self.onion_count.load(Ordering::Relaxed),
                clearnet_pages: self.clearnet_count.load(Ordering::Relaxed),
                crawl_duration: self.format_elapsed(),
                pages_per_second: if elapsed > 0.0 { total as f64 / elapsed } else { 0.0 },
                total_data_downloaded_mb: self.total_size_bytes.load(Ordering::Relaxed) as f64 / (1024.0 * 1024.0),
                total_links_found: self.total_links_found.load(Ordering::Relaxed),
            },
            top_domains,
            status_code_distribution: status_dist.into_iter().collect(),
            content_type_distribution: ct_dist.into_iter().collect(),
            error_distribution: error_dist.into_iter().collect(),
            largest_pages: largest.into_iter().map(|(u, s)| LargestPage { url: u, size_bytes: s }).collect(),
        }
    }

    pub fn save_report(&self, path: &str) -> std::io::Result<()> {
        let report = self.generate_report();
        let json = serde_json::to_string_pretty(&report)?;
        std::fs::write(path, json)?;
        Ok(())
    }

    pub fn print_summary(&self) {
        let report = self.generate_report();
        let s = &report.summary;

        println!("\n{}", "=".repeat(60));
        println!("  {}", "AGGREGATION REPORT".cyan().bold());
        println!("{}", "=".repeat(60));

        println!("\n  {}", "Overview".bold());
        println!("  ├─ Pages crawled:    {}", format!("{}", s.total_pages_crawled).green());
        println!("  ├─ Unique domains:   {}", format!("{}", s.total_unique_domains).blue());
        println!("  ├─ Clearnet pages:   {}", s.clearnet_pages);
        println!("  ├─ Onion pages:      {}", format!("{}", s.onion_pages).magenta());
        println!("  ├─ Errors:           {}", format!("{}", s.total_errors).red());
        println!("  ├─ Data downloaded:  {:.1} MB", s.total_data_downloaded_mb);
        println!("  ├─ Links found:      {}", s.total_links_found);
        println!("  ├─ Duration:         {}", s.crawl_duration);
        println!("  └─ Speed:            {:.1} pages/sec", s.pages_per_second);

        // Status codes
        if !report.status_code_distribution.is_empty() {
            println!("\n  {}", "Status Codes".bold());
            for (code, count) in &report.status_code_distribution {
                let code_str = format!("{}", code);
                let colored = if *code >= 200 && *code < 300 {
                    code_str.green()
                } else if *code >= 300 && *code < 400 {
                    code_str.yellow()
                } else {
                    code_str.red()
                };
                println!("    {}: {}", colored, count);
            }
        }

        // Top domains
        if !report.top_domains.is_empty() {
            println!("\n  {}", "Top Domains (by pages)".bold());
            for (i, (domain, count)) in report.top_domains.iter().take(15).enumerate() {
                let marker = if domain.contains(".onion") {
                    "[TOR] ".magenta().to_string()
                } else {
                    String::new()
                };
                println!(
                    "    {:>3}. {}{} — {} pages",
                    (i + 1).to_string().dimmed(),
                    marker,
                    domain,
                    format!("{}", count).cyan()
                );
            }
        }

        // Content types
        if !report.content_type_distribution.is_empty() {
            println!("\n  {}", "Content Types".bold());
            for (ct, count) in &report.content_type_distribution {
                println!("    {} {}: {}", "•".dimmed(), ct, count);
            }
        }

        // Errors
        if !report.error_distribution.is_empty() {
            println!("\n  {}", "Error Types".bold());
            for (err, count) in &report.error_distribution {
                println!("    {} {}: {}", "•".red(), err, count);
            }
        }

        // Largest
        if !report.largest_pages.is_empty() {
            println!("\n  {}", "Largest Pages".bold());
            for p in report.largest_pages.iter().take(5) {
                let kb = p.size_bytes as f64 / 1024.0;
                let url_short = if p.url.len() > 80 { &p.url[..80] } else { &p.url };
                println!("    {} {:.1} KB — {}", "•".dimmed(), kb, url_short);
            }
        }

        println!("\n{}\n", "=".repeat(60));
    }
}

// ─── Serializable Report Structs ────────────────────────────────────────────────

#[derive(Serialize)]
pub struct Report {
    pub summary: ReportSummary,
    pub top_domains: Vec<(String, usize)>,
    pub status_code_distribution: Vec<(u16, usize)>,
    pub content_type_distribution: Vec<(String, usize)>,
    pub error_distribution: Vec<(String, usize)>,
    pub largest_pages: Vec<LargestPage>,
}

#[derive(Serialize)]
pub struct ReportSummary {
    pub total_pages_crawled: usize,
    pub total_unique_domains: usize,
    pub total_errors: usize,
    pub onion_pages: usize,
    pub clearnet_pages: usize,
    pub crawl_duration: String,
    pub pages_per_second: f64,
    pub total_data_downloaded_mb: f64,
    pub total_links_found: usize,
}

#[derive(Serialize)]
pub struct LargestPage {
    pub url: String,
    pub size_bytes: u64,
}

File: xue/aggregator.py
import collections
import heapq
import json
import threading
import time
from datetime import datetime, timedelta
from urllib.parse import urlparse

from xue.ansi import C
from xue.classifier import DomainClassifier


class Aggregator:
    def __init__(self):
        self._lock = threading.Lock()
        self.domains = collections.Counter()
        self.domain_first_seen = {}
        self.status_codes = collections.Counter()
        self.content_types = collections.Counter()
        self.total_links_found = 0
        self.total_size_bytes = 0
        self.errors = collections.Counter()
        self.error_urls = collections.deque(maxlen=200)
        self.onion_count = 0
        self.clearnet_count = 0
        self.classifier = DomainClassifier()
        self.domain_categories = collections.defaultdict(list)
        self.domain_to_category = {}
        self.start_time = time.time()
        self.largest_pages = []
        self.robots_blocked = 0
        self.retry_after_count = 0
        self.api_endpoints_found = 0
        self.broken_links = []
        self.duplicate_pages = 0
        self.tech_findings: dict[str, list[dict]] = {}
        self.content_stats = {"pages_extracted": 0, "total_words": 0, "total_chars": 0}
        self.seo_reports: list[dict] = []
        self.budget_exceeded_domains: list[str] = []

    def record_page(self, url, status_code, content_type, links_found, depth, size_bytes=0, title=""):
        with self._lock:
            parsed = urlparse(url)
            domain = parsed.netloc
            self.domains[domain] += 1
            if domain not in self.domain_first_seen:
                self.domain_first_seen[domain] = datetime.now().isoformat()
            if domain not in self.domain_to_category:
                cat = self.classifier.classify(domain, title)
                self.domain_to_category[domain] = cat
                self.domain_categories[cat].append(domain)
            self.status_codes[status_code] += 1
            ct = (content_type or "unknown").split(";")[0].strip()
            self.content_types[ct] += 1
            self.total_links_found += links_found
            self.total_size_bytes += size_bytes
            if domain.endswith(".onion"):
                self.onion_count += 1
            else:
                self.clearnet_count += 1
            if len(self.largest_pages) < 10:
                heapq.heappush(self.largest_pages, (size_bytes, url))
            else:
                heapq.heappushpop(self.largest_pages, (size_bytes, url))
            if status_code >= 400:
                self.broken_links.append({"url": url, "status": status_code, "source": ""})
                if len(self.broken_links) > 500:
                    self.broken_links = self.broken_links[-500:]
            if "application/json" in ct:
                self.api_endpoints_found += 1

    def record_error(self, url, error_msg):
        with self._lock:
            error_type = type(error_msg).__name__ if not isinstance(error_msg, str) else error_msg.split(":")[0]
            self.errors[error_type] += 1
            self.error_urls.append((url, str(error_msg)))

    def record_robots_blocked(self):
        with self._lock:
            self.robots_blocked += 1

    def record_retry_after(self):
        with self._lock:
            self.retry_after_count += 1

    def record_duplicate(self):
        with self._lock:
            self.duplicate_pages += 1

    def record_tech_findings(self, url: str, findings: list):
        with self._lock:
            domain = urlparse(url).netloc
            if domain not in self.tech_findings:
                self.tech_findings[domain] = []
            for f in findings:
                self.tech_findings[domain].append({
                    "name": f.name,
                    "confidence": f.confidence,
                    "evidence": f.evidence,
                })

    def record_content(self, url: str, extracted):
        with self._lock:
            self.content_stats["pages_extracted"] += 1
            self.content_stats["total_words"] += extracted.word_count
            self.content_stats["total_chars"] += extracted.char_count

    def record_seo(self, report):
        with self._lock:
            self.seo_reports.append({
                "url": report.url,
                "title": report.title,
                "title_length": report.title_length,
                "meta_description": report.meta_description,
                "h1_count": report.h1_count,
                "has_canonical": report.has_canonical,
                "images_without_alt": report.images_without_alt,
                "issues": report.heading_issues,
            })

    def record_budget_exceeded(self, url: str):
        with self._lock:
            domain = urlparse(url).netloc
            if domain not in self.budget_exceeded_domains:
                self.budget_exceeded_domains.append(domain)

    def get_live_stats(self):
        with self._lock:
            elapsed = time.time() - self.start_time
            total = sum(self.status_codes.values())
            rate = total / elapsed if elapsed > 0 else 0
            return {
                "total_pages": total,
                "total_domains": len(self.domains),
                "total_errors": sum(self.errors.values()),
                "onion_pages": self.onion_count,
                "clearnet_pages": self.clearnet_count,
                "pages_per_sec": round(rate, 2),
                "elapsed": str(timedelta(seconds=int(elapsed))),
                "total_size_mb": round(self.total_size_bytes / (1024 * 1024), 1),
            }

    def generate_report(self):
        with self._lock:
            elapsed = time.time() - self.start_time
            total = sum(self.status_codes.values())
            return {
                "summary": {
                    "total_pages_crawled": total,
                    "total_unique_domains": len(self.domains),
                    "total_errors": sum(self.errors.values()),
                    "robots_blocked": self.robots_blocked,
                    "retry_after_count": self.retry_after_count,
                    "duplicate_pages": self.duplicate_pages,
                    "api_endpoints_found": self.api_endpoints_found,
                    "broken_links_count": len(self.broken_links),
                    "onion_pages": self.onion_count,
                    "clearnet_pages": self.clearnet_count,
                    "crawl_duration": str(timedelta(seconds=int(elapsed))),
                    "pages_per_second": round(total / elapsed, 2) if elapsed > 0 else 0,
                    "total_data_downloaded_mb": round(self.total_size_bytes / (1024 * 1024), 1),
                    "total_links_found": self.total_links_found,
                },
                "top_domains": self.domains.most_common(30),
                "domains_by_category": {cat: doms for cat, doms in sorted(self.domain_categories.items(), key=lambda x: len(x[1]), reverse=True)},
                "status_code_distribution": dict(self.status_codes.most_common()),
                "content_type_distribution": dict(self.content_types.most_common()),
                "error_distribution": dict(self.errors.most_common()),
                "top_error_urls": list(self.error_urls)[-20:],
                "largest_pages": sorted(self.largest_pages, key=lambda x: x[0], reverse=True),
                "broken_links": self.broken_links[-50:],
            }

    def save_report(self, filepath):
        report = self.generate_report()
        with open(filepath, "w", encoding="utf-8") as f:
            json.dump(report, f, indent=2, ensure_ascii=False, default=str)
        return filepath

    def print_summary(self):
        report = self.generate_report()
        s = report["summary"]
        print(f"\n{'='*60}")
        print(f"{C.CYAN}{C.BOLD}  AGGREGATION REPORT{C.RST}")
        print(f"{'='*60}")
        print(f"\n  {C.BOLD}Overview{C.RST}")
        print(f"  ├─ Pages crawled:      {C.GREEN}{s['total_pages_crawled']}{C.RST}")
        print(f"  ├─ Unique domains:     {C.BLUE}{s['total_unique_domains']}{C.RST}")
        print(f"  ├─ Clearnet pages:     {C.WHITE}{s['clearnet_pages']}{C.RST}")
        print(f"  ├─ Onion pages:        {C.MAGENTA}{s['onion_pages']}{C.RST}")
        print(f"  ├─ Errors:             {C.RED}{s['total_errors']}{C.RST}")
        if s.get('robots_blocked', 0) > 0:
            print(f"  ├─ Robots blocked:     {C.YELLOW}{s['robots_blocked']}{C.RST}")
        if s.get('retry_after_count', 0) > 0:
            print(f"  ├─ Retry-After hits:   {C.YELLOW}{s['retry_after_count']}{C.RST}")
        if s.get('duplicate_pages', 0) > 0:
            print(f"  ├─ Duplicate pages:    {C.DIM}{s['duplicate_pages']}{C.RST}")
        if s.get('api_endpoints_found', 0) > 0:
            print(f"  ├─ API endpoints:      {C.CYAN}{s['api_endpoints_found']}{C.RST}")
        if s.get('broken_links_count', 0) > 0:
            print(f"  ├─ Broken links:       {C.RED}{s['broken_links_count']}{C.RST}")
        print(f"  ├─ Data downloaded:    {s['total_data_downloaded_mb']} MB")
        print(f"  ├─ Links found:        {s['total_links_found']}")
        print(f"  ├─ Duration:           {s['crawl_duration']}")
        print(f"  └─ Speed:              {s['pages_per_second']} pages/sec")
        if report["status_code_distribution"]:
            print(f"\n  {C.BOLD}Status Codes{C.RST}")
            for code, count in sorted(report["status_code_distribution"].items()):
                color = C.GREEN if 200 <= code < 300 else C.YELLOW if 300 <= code < 400 else C.RED
                print(f"    {color}{code}{C.RST}: {count}")
        if report["top_domains"]:
            print(f"\n  {C.BOLD}Top Domains (by pages){C.RST}")
            for i, (domain, count) in enumerate(report["top_domains"][:15], 1):
                marker = f"{C.MAGENTA}[TOR]{C.RST} " if domain.endswith(".onion") else ""
                print(f"    {C.DIM}{i:>3}.{C.RST} {marker}{domain} — {C.CYAN}{count}{C.RST} pages")
        if report["domains_by_category"]:
            print(f"\n  {C.BOLD}Domains by Category{C.RST}")
            cat_colors = {
                "Pornography / Adult": C.RED, "Social Media": C.BLUE, "Gaming": C.GREEN,
                "Game Store / Marketplace": C.GREEN, "Technology": C.CYAN, "News / Media": C.YELLOW,
                "Forums / Community": C.MAGENTA, "Shopping / E-Commerce": C.YELLOW,
                "Streaming / Entertainment": C.MAGENTA, "Education": C.CYAN, "Government": C.WHITE,
                "Finance / Crypto": C.GREEN, "AI / Machine Learning": C.CYAN,
                "Search Engine": C.BLUE, "Advertising / Tracking": C.GRAY, "Uncategorized": C.DIM,
            }
            for cat, domains in report["domains_by_category"].items():
                color = cat_colors.get(cat, C.DIM)
                print(f"\n    {color}{C.BOLD}{cat}{C.RST} ({len(domains)} domains)")
                for d in domains[:8]:
                    print(f"      {C.DIM}•{C.RST} {d}")
                if len(domains) > 8:
                    print(f"      {C.DIM}  ... and {len(domains) - 8} more{C.RST}")
        if report["content_type_distribution"]:
            print(f"\n  {C.BOLD}Content Types{C.RST}")
            for ct, count in list(report["content_type_distribution"].items())[:10]:
                print(f"    {C.DIM}•{C.RST} {ct}: {count}")
        if report["error_distribution"]:
            print(f"\n  {C.BOLD}Error Types{C.RST}")
            for err, count in list(report["error_distribution"].items())[:10]:
                print(f"    {C.RED}•{C.RST} {err}: {count}")
        if report.get("broken_links"):
            print(f"\n  {C.BOLD}Broken Links (4xx/5xx){C.RST}")
            for bl in report["broken_links"][:10]:
                print(f"    {C.RED}{bl['status']}{C.RST} {bl['url'][:80]}")
        if report["largest_pages"]:
            print(f"\n  {C.BOLD}Largest Pages{C.RST}")
            for size, url in report["largest_pages"][:5]:
                size_kb = size / 1024 if size else 0
                print(f"    {C.DIM}•{C.RST} {size_kb:.1f} KB — {url[:80]}")
        print(f"\n{'='*60}\n")
